### PR TITLE
Only wait for th elids that you are interested in.

### PR DESCRIPTION
**Pull request by baldersheim**
@toregge PR
Just for discussion.

### CODE REVIEW

**Comment by toregge** [comment]
Assuming a large time interval between operations on the same lid (document), we have 4 states:
A) steady state, nothing happened
B) feed operation in progress, main thread or attribute/indexing/summary writer worker threads are processing feed operation
C) feed operation done, but commit not started
D) feed operation done, commit started, but not completed.

With visibility delay being 0, only state A and B are relevant.
With nonzero visibility delay, states A, B, C and D are relevant.

Removing the first assumption leads to more states (superpositioning of the above states).

The commit above fails to track state B when visibility delay is zero (uses nooplidtracker) and fails to track state C and D when visibilitydelay is nonzero.

File: build_settings.cmake [hunk 1]
```diff
@@ -73,7 +73,7 @@ else()
 endif()
 
 # C and C++ compiler flags
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -O3 -fno-omit-frame-pointer ${C_WARN_OPTS} -fPIC ${VESPA_CXX_ABI_FLAGS} ${VESPA_XXHASH_DEFINE} -DBOOST_DISABLE_ASSERTS ${VESPA_CPU_ARCH_FLAGS} ${EXTRA_C_FLAGS}")
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -O0 -fno-omit-frame-pointer ${C_WARN_OPTS} -fPIC ${VESPA_CXX_ABI_FLAGS} ${VESPA_XXHASH_DEFINE} -DBOOST_DISABLE_ASSERTS ${VESPA_CPU_ARCH_FLAGS} ${EXTRA_C_FLAGS}")
```
Comment (toregge) [inline build_settings.cmake:76]: Consider reverting the above change.

Comment (baldersheim) [reply build_settings.cmake]: Consider is a bit on the weak side :)

**Comment by baldersheim** [comment]
There is more coming

**Comment by baldersheim** [comment]
@toregge Now you can take another look. Tests are added and interfaces refactored to only need 1 lookup in the fast path.

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.cpp [hunk 2]
```diff
@@ -58,12 +90,137 @@ PendingLidTracker::consume(uint32_t lid) {
...
+        if (entry.second.need_commit) {
+            toCommit.emplace_back(entry.first);
+            entry.second.inflight_commit ++;
+            entry.second.need_commit = false;
+        }
+    }
+    return std::make_unique<common::internal::CommitList>(std::move(toCommit), *this);
+}
+
+}
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.cpp:226]: Newline is missing at end of file.

**Comment by toregge** [comment]
@geirst : FYI

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.cpp (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.cpp:226]: Fixed

**Review by toregge — APPROVED** [review]

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h [hunk 3]
```diff
@@ -30,33 +32,102 @@ class IPendingLidTracker {
...
+    void waitComplete() const;
 private:
-    void consume(uint32_t ) override { }
+    virtual State waitState(State state, uint32_t lid) const = 0;
+    virtual State waitState(State state, const LidList & lids) const = 0;
+    virtual State waitState(State state) const = 0;
+};
+
+/**
+ * Base class for doing 2 phase lidtracking.
```
[the remark this answers is not in the document]
Comment (geirst) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h:60]: Also describe the difference between a Token and a Snapshot.

File: searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp [hunk 4]
```diff
@@ -585,19 +605,20 @@ StoreOnlyFeedView::internalRemove(FeedToken token, const RemoveOperationWithDocI
          rmOp.getSubDbId(), rmOp.getLid(), rmOp.getPrevSubDbId(), rmOp.getPrevLid(), _params._subDbId);
 
     PendingNotifyRemoveDone pendingNotifyRemoveDone = adjustMetaStore(rmOp, docId.getGlobalId(), docId);
+    auto uncommitted = _pendingLidsForCommit->produce(rmOp.getLid());
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp:608]: This is named commitToken other places in this file. Any reason we use another name here? If not, consider consolidating.

Comment (geirst) [reply searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp:608]: Note, there are more places we use _uncomitted_ in this file.

**Review by geirst — COMMENTED** [review]
Overall looks good. I would like to see more descriptions of the new concepts added.

File: searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp (hunk 4 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp:608]: Unified on uncommitted.

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h [hunk 5]
```diff
@@ -5,9 +5,11 @@
 #include <vespa/vespalib/stllike/hash_map.h>
 #include <mutex>
 #include <condition_variable>
+#include <vector>
 
 namespace proton {
 
+/** Interface for tracking lids in the feed pipeline */
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h]: Done

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h [hunk 6]
```diff
@@ -30,33 +32,102 @@ class IPendingLidTracker {
...
+    State waitFor(MonitorGuard & guard, State state, uint32_t lid) const override;
+
     vespalib::hash_map<uint32_t, uint32_t> _pending;
 };
 
+namespace common::internal {
+    class CommitList;
+}
+/**
+ * Use for tracking lids in 2 phases.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h]: Done

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h [hunk 7]
```diff
@@ -30,33 +32,102 @@ class IPendingLidTracker {
...
+    virtual LidList pendingLids() const = 0;
+    virtual State waitFor(MonitorGuard & guard, State state, uint32_t lid) const = 0;
+    MonitorGuard getGuard() { return MonitorGuard(_mutex); }
+    mutable std::mutex                     _mutex;
+    mutable std::condition_variable        _cond;
 };
 
-class PendingLidTracker : public IPendingLidTracker {
+/**
+ * Use for tracking lids in a single phase.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h]: Done

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h]: Done

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h [hunk 8]
```diff
@@ -30,33 +32,102 @@ class IPendingLidTracker {
...
 private:
     virtual void consume(uint32_t lid) = 0;
-    std::mutex _mutex;
-    std::condition_variable _cond;
-    vespalib::hash_map<uint32_t, uint32_t> _pending;
 };
 
-class NoopLidTracker : public IPendingLidTracker {
+/**
+ * This is an interface for checking/waiting the state of a lid in the feedpipeline
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h]: Done

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h [hunk 9]
```diff
@@ -84,7 +97,8 @@ class PendingLidTrackerBase : public IPendingLidTracker,
 };
 
 /**
- * Use for tracking lids in a single phase.
+ * Use for tracking lids whenn visibility-delay is zero and commit is implicit.
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h:100]: whenn -> when

File: searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp [hunk 10]
```diff
@@ -318,7 +318,7 @@ StoreOnlyFeedView::internalPut(FeedToken token, const PutOperation &putOp)
          _params._subDbId, doc->toString(true).size(), doc->toString(true).c_str());
 
     PendingNotifyRemoveDone pendingNotifyRemoveDone = adjustMetaStore(putOp, docId.getGlobalId(), docId);
-    auto commitToken = _pendingLidsForCommit->produce(putOp.getLid());
+    auto unconmitted = _pendingLidsForCommit->produce(putOp.getLid());
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp:321]: unconmitted -> uncommitted

**Review by geirst — APPROVED** [review]
👍

File: searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h (hunk 9 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/pendinglidtracker.h]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp (hunk 10 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/storeonlyfeedview.cpp]: Fixed

**Pull request merged**